src/sphmultipcqueue.h

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright (c) 2016 IBM Corporation.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ * 
+ * Contributors:
+ *     IBM Corporation, Paul Clarke - based heavily on sphsinglepcqueue
+ */
+
+#ifndef __SPH_MULTI_PC_QUEUE_H
+#define __SPH_MULTI_PC_QUEUE_H
+
+/**! \file sphmultipcqueue.h
+*  \brief Shared Persistent Heap, multi producer multi consumer queue.
+*	For shared memory multi-thread/multi-core applications.
+*	This implementation uses transactional memory operations to
+*	implement Lock Free Producer/Consumer queues (SPHMultiPCQueue_t).
+*
+*	This API supports atomic allocation of storage for queue
+*	entries for zero copy persistence and sharing. Zero copy queues
+*	divides the process of producing a queue entry in to three steps:
+*	- Allocate the queue entry (and initialize the header)
+*	- Use the return entry handle to fill in application specific data.
+*	- Marks the entry complete in the header.
+*
+*	\code
+#include <sphlfentry.h>
+SPHLFEntryHandle_t *handle;
+typedef struct data_struct11 {
+	double	field0;
+	int		field1;
+	int		field2;
+	void	*field3;
+} data_struct11;
+
+// Allocate zero copy queue entry
+handle = SPHMPMCQMultiAllocStrideDirectTM (pcqueue);
+if (handle)
+{	// insert data into the allocated queue entry
+	sphLFEntryID_t tmpl;
+	data_struct11 *struct_ptr;
+
+	tmpl = SPHMPMCQGetEntryTemplate(pcq);
+
+	struct_ptr = (data_struct11 *)SPHLFEntryDirectGetFreePtr(handle);
+	if (struct_ptr)
+	{	// store struct fields directly into allocated queue entry
+		struct_ptr->field0 = data_double1;
+		struct_ptr->field1 = data_int2;
+		struct_ptr->field2 = data_int3;
+		struct_ptr->field3 = (void*)sas_data_buff2;
+	} else {
+		printf("SPHENTRYALLOCSTRUCT(%p, data_struct11) failed)\n",
+			   handle);
+	}
+	// Mark the entry complete and available to the consumer
+	SPHLFEntryDirectComplete(handle)
+} else {
+	while (SPHMPMCQMultiIsQueueFull(pcqueue))
+	{
+		// pacing code
+	}
+}
+*	\endcode
+*
+*	The consumer can access queue entries once they are marked complete.
+*	The consumer:
+*	- checks (spins) for the next allocated entry to become complete
+*	- uses the returned entry handle to directly access the entry contents
+*	- When done processing the queue entry, it marks the entry header
+*	invalid and deallocates the entry
+*	- This makes the next queue entry available, if any
+*
+*	\code
+#include <sphlfentry.h>
+SPHLFEntryHandle_t *handle;
+typedef struct data_struct11 {
+	double	field0;
+	int	field1;
+	int	field2;
+	void	*field3;
+} data_struct11;
+
+// Get next queue entry
+handle = SPHMPMCQMultiGetNextCompleteDirectTM (pcqueue, &handle0);
```

Comment:
there are a couple of issues with this line, so I'm glad you pointed it out.  The API is wrong (MPMCQMulti) and the correct API (SPHMPMCQGetNextCompleteDirectTM) doesn't have that 2nd parameter, anyway.  It should be:
handle = SPHMPMCQGetNextCompleteDirectTM(pcqueue);